ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -1148,26 +1152,60 @@ public List<AmbryReplica> getReplicaIdsByState(AmbryPartition partition, Replica
       Set<AmbryReplica> replicas = new HashSet<>();
       List<String> dcs = dcName != null ? Collections.singletonList(dcName)
           : dcToDcInfo.values().stream().map(dcInfo -> dcInfo.dcName).collect(Collectors.toList());
+      String partitionName = partition.toPathString();
       for (String dc : dcs) {
-        String resourceName = partitionToResourceNameByDc.get(dc).get(partition.toPathString());
+        int replicaSizeBefore = replicas.size();
+        String resourceName = partitionToResourceNameByDc.get(dc).get(partitionName);
         RoutingTableSnapshot routingTableSnapshot =
             clusterMapConfig.clusterMapUseAggregatedView ? globalRoutingTableSnapshotRef.get()
                 : dcToRoutingTableSnapshotRef.get(dc).get();
-        routingTableSnapshot.getInstancesForResource(resourceName, partition.toPathString(), state.name())
-            .stream()
-            .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
-            .filter(dataNode -> dataNode.getDatacenterName().equals(dc))
-            .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
-            .filter(Objects::nonNull)
-            .forEach(replicas::add);
+        getReplicaIdsByStateInRoutingTableSnapshot(routingTableSnapshot, dc, resourceName, partitionName, state,
+            replicas);
+        if (replicas.size() != replicaSizeBefore) {
+          // We found the replicas from the resource name, just continue to next dc.
+          continue;
+        }
+        // There is some change made to resources in IDEALSTATES and it's not yet reflected in EXTERNAALVIEW.
+        // Try to search the partition in all resources
+        helixClusterManagerMetrics.resourceNameMismatchCount.inc();
+        String resourceNameInCache =
+            dcToPartitionNameToResourceCache.computeIfAbsent(dc, k -> new ConcurrentHashMap<>())
+                .get(partition.toPathString());
+        if (resourceNameInCache == null) {
+          helixClusterManagerMetrics.resourceNameCacheMissCount.inc();
+          Collection<String> resourcesFromRoutingTable = routingTableSnapshot.getResources();
+          for (String resource : resourcesFromRoutingTable) {
+            getReplicaIdsByStateInRoutingTableSnapshot(routingTableSnapshot, dc, resource, partitionName, state,
+                replicas);
+            if (replicas.size() != replicaSizeBefore) {
+              // We found the partition in this resource, put this resource in the case.
+              dcToPartitionNameToResourceCache.get(dc).putIfAbsent(partitionName, resource);
+              break;
+            }
+          }
+        } else {
+          helixClusterManagerMetrics.resourceNameCacheHitCount.inc();
+          getReplicaIdsByStateInRoutingTableSnapshot(routingTableSnapshot, dc, resourceNameInCache, partitionName,
```

Comment:
Just curious, after the move to full-auto is complete, would this case happen that a partition id is moved from one resource to another?

Reply:
I updated the PR to reflect the case where a partition is moved to a different resource. Basically we want to allow multiple resources for any given partition as long as the instance list is different.